Handle missing data fields in UI

AlexAxthelm opened the pull request:
Closes #255 

Adds a "None" badge to fields in the presence of absent data (No temperature target defined, for example), and adds corresponding filter option in dropdown filter menus when needed

Note that the "None" Badges do not have clarifying tooltips yet, that is tracked as #384

AlexAxthelm committed:
chore(filters): add ABSENT sentinel + helpers to support virtual “None” values

Adds ABSENT symbol, isAbsent, coalesceOptional, and filter token helpers.

Includes Vitest unit tests.

No runtime behavior changes yet.

AlexAxthelm committed:
chore(filters): add facet option builder with ABSENT bucket

Adds a tiny utility to build {label,value} options from raw values, injecting a None option (__ABSENT__) when appropriate.

Includes Vitest unit tests.

No UI/components changed yet.

AlexAxthelm committed:
chore(normalize): add helper to coalesce optional fields to ABSENT

* Adds normalizeOptionalFields (+ list variant) to consistently map
  undefined|null → ABSENT for selected keys.
* Includes Vitest tests.

AlexAxthelm committed:
chore(facets): add missing-aware matcher to pair with options builder

AlexAxthelm committed:
fix(search filter): handle optional values in search dropdown

Handle optional values in the search dropdown for temperature,
extendable to other dropdowns.

Closes #255

AlexAxthelm committed:
Merge branch 'main' into fix/255-optional-fields

AlexAxthelm committed:
fix(search) use facet builder for all dropdowns

This extends the auto-inclusion of a "virtual None" option for absent
values to all dropdowns in the SearchSection

AlexAxthelm committed:
fix(search): use facets to filter pathwayType and modelYearEnd

AlexAxthelm committed:
fix(search): use facets to filter Sectors

Sectors don't take a simple array (array of `object`), so slightly more
invovled, and worth splitting into their own commit

AlexAxthelm committed:
fix(search): use facets to filter Geography

Updating the construction of geography filter options to allow for None
option

AlexAxthelm committed:
test(search): test for matching absent value (none)

AlexAxthelm committed:
build(deps): add @testing-library/user-event

Simulate User events in integration tests.

AlexAxthelm committed:
test(integration): HomePage renders ‘None’ options in filters

Testing Sector and Geography because they aren't simple arrays, and
Temperature because it is.

AlexAxthelm committed:
extend integration test

also address some lints and format

AlexAxthelm committed:
chore(badge): Add absent-aware badge (`BadgeMaybeAbsent`)

AlexAxthelm committed:
fix(badges): Use absent-aware badges in UI

Closes #255

AlexAxthelm committed:
linting: default string coercion

AlexAxthelm committed:
test(badge): Test for "None" in presence of absent value

jdhoffa commented on src/utils/facets.test.ts:
```diff
@@ -0,0 +1,146 @@
+} from "./facets";
+import { ABSENT_FILTER_TOKEN } from "./absent";
+
+import { ABSENT_FILTER_TOKEN } from "./absent";
```
Agree

AlexAxthelm replied:
https://github.com/RMI/pbtar/pull/374/commits/bd26bac962ac62ed92eb5090b2e8130a74048c1e

jdhoffa reviewed (changes requested):
Apart from the erroneous double import in `facets.test.ts` flagged by CoPilot, this LGTM. Re-request review once that is merged and I will approve. 

---
This is quite a large PR, so I'd just like to re-iterate my understanding of what is happening here. The effect of these changes is clear, so I will delve mainly into implementation details.

Fundamentally, the addition here is of a "None" badge to display in situations where optional data fields are missing (temperature target, geography, sector).

- Gains an `ABSENT` sentinel, a `unique symbol` that can't collide with or be confused as real data, along with helpers to work with this sentinel (and an `

- A new  functional component, `BadgeMaybeAbsent`, is added which automatically renders a "None" label in case of absent values, and otherwise shows values with optional formatting.
- Updates usage of `Badge` -> `BadgeMaybeAbsent` in relevant places

---

- `buildOptionsFromValues` function to build out filtering options, deduping and ordering them; as well as mapping various missing value types `undefined / null` to the `ABSENT` sentinel. Creates an array of `label, value` options, and appendsa `"None", value: "__ABSENT__" in case absent values are found
- `hasAbsent` - Checks if the `ABSENT` sentinel is found, as above
- `withAbsentOption` - Appends a "None" to the dropdown menu, as above

AlexAxthelm committed:
Merge branch 'main' into fix/255-optional-fields

AlexAxthelm committed:
Remove duplicate import

AlexAxthelm committed:
update test strings

Not strictly necessary, but useful since these are testing for
`not.toBeInTheDocument()`, so having the wrong strings here is a bad
test (they would never be in the document).

AlexAxthelm commented on src/pages/HomePage.test.tsx:
```diff
@@ -51,3 +52,151 @@ describe("HomePage component", () => {
+      "Scenario C (empty sectors[], empty geo[], 1.5°C)",
+    ]);
+    expectHidden([
+      "Scenario B (Power, EU, 2°C)",
```
https://github.com/RMI/pbtar/pull/374/commits/dc427bd1810ccac96e3593a6ac5593707314b7f8

AlexAxthelm commented on src/pages/HomePage.test.tsx:
```diff
@@ -51,3 +52,151 @@ describe("HomePage component", () => {
+      "Scenario C (empty sectors[], empty geo[], 1.5°C)",
+    ]);
+    expectHidden([
+      "Scenario B (Power, EU, 2°C)",
```
https://github.com/RMI/pbtar/pull/374/commits/dc427bd1810ccac96e3593a6ac5593707314b7f8
